### PR TITLE
Move custom schedule tools away from using PeriodicTasks in favor of the new schedule framework

### PR DESCRIPTION
**Pull request by stephherbers**
https://github.com/dimagi/open-chat-studio/pull/471#:~:text=Work%20in%20Progress-,%23437,-migrate%20tools%20off

Incorporated feedback from previous [draft PR ](https://github.com/dimagi/open-chat-studio/pull/471) such that custom_schedule_params takes precedence over the action which is possible after the action was made option in a previous PR

### CODE REVIEW

File: apps/chat/agent/tools.py
Comment (snopoke) [inline apps/chat/agent/tools.py:128]: should this function return errors if it's not valid

File: apps/chat/agent/tools.py
Comment (snopoke) [inline apps/chat/agent/tools.py:132]: nit: you can use the experiment from the session `experiment_session.experiment`

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:84]: should check if the experiment exists and ignore tasks if not

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:52]: should we bother copying expired schedules?

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:64]: also wondering if we need to copy over schedules that have no more repetitions

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:37]: If the task never expires doesn't that imply an infinite number of repetitions? I don't think we support that but maybe we should?

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:103]: Why do we use an action here instead of `custom_schedule_params`? I think all of these are use initiated schedules.

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:80]: Why do we filter by `enabled=False`? Shouldn't it be `enabled=True`?

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:119]: nit: make it elidable
```suggestion
    migrations.RunPython(migrate_periodic_tasks, elidable=True),
```

**Review by snopoke — COMMENTED** [review]
Generally looks good. Mostly questions about the migration.

File: apps/events/models.py
Comment (SmittieC) [inline apps/events/models.py:309]: nit: make this a `@cached_property` as well

File: apps/chat/agent/tools.py
Comment (SmittieC) [reply apps/chat/agent/tools.py:128]: Since the return value of this will be fed back to the agent, I think we should catch any error and return something like "Could not do that". The LLM will use this to tell the user that it couldn't create a scheduled message

File: apps/chat/agent/tools.py
Comment (SmittieC) [inline apps/chat/agent/tools.py:126]: Looks like `cleaned_data["name"]` will fail the form validation, since it would be missing? i don't think the name here is too important, so we could probably generate it. Also would be a good idea to add a test for this function.

Comment (stephherbers) [reply apps/chat/agent/tools.py:127]: [7ec855d](https://github.com/dimagi/open-chat-studio/pull/527/commits/7ec855d043b90502bb536237c2e585b95a38fdd7) modified so it gernerateds a name with a unique uuid

File: apps/chat/agent/tools.py
Comment (stephherbers) [reply apps/chat/agent/tools.py:129]: is this better? [e3cc6b2](https://github.com/dimagi/open-chat-studio/pull/527/commits/e3cc6b2779fcef5bf56c86ee140ecc359b3516c5)

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (stephherbers) [reply apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:37]: ah I didn't think about that but that makes sense. In that case how should remaining time be represented?

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (SmittieC) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:6]: these model imports can be removed

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (SmittieC) [inline apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:77]: Ooh, careful here. `datetime.now()` returns a naive date, which is incompatable with timzone aware dates, so you will not be able to do for instance `task.expires - current_time`.

**Review by SmittieC — COMMENTED** [review]
Just two comments, otherwise looks great! This has been a bit more complicated than I initially thought it would have been.

Have you tested the migration run yet?

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (snopoke) [reply apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py:77]: You can use `timezone.now()` (`from django.utils import timezone`)

**Review by snopoke — COMMENTED** [review]
Just a general comment for another PR; with better tool descriptions I think we should be able to combine these two tools into one e.g. by including examples.

File: apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py
Comment (stephherbers) [reply apps/events/migrations/0014_move_schedule_messages_off_periodic_tasks.py]: [b2d92d1](https://github.com/dimagi/open-chat-studio/pull/527/commits/b2d92d17f732edd2bb9a9e62929eef7614ea1d16)

File: apps/events/models.py
Comment (stephherbers) [reply apps/events/models.py:310]: ah yes thank you

**Comment by stephherbers** [comment]
@SmittieC I did test the migration locally, but if you have suggestions for testing on prod data lmk

**Comment by SmittieC** [comment]
> @SmittieC I did test the migration locally, but if you have suggestions for testing on prod data lmk

I DM'd you on Slack on the process that I normally follow. Let me know if you have any questions

**Comment by snopoke** [comment]
@stephherbers other than the migration testing I think this is OK to merge and my remaining comments can be addressed in a follow up PR if you'd prefer to do it that way.

**Comment by stephherbers** [comment]
@snopoke @SmittieC migration tested and I agree this can be merged to unblock Simon and address the comments separately after approval

**Comment by SmittieC** [comment]
> @snopoke @SmittieC migration tested and I agree this can be merged to unblock Simon and address the comments separately after approval

Thanks Steph, this is great. I took the liberty to fix the conflicting migration files, hope you don't mind.

**Review by SmittieC — APPROVED** [review]

**Pull request merged**
